### PR TITLE
[test] Add CPU affinity test

### LINKED ISSUE #100: Add CPU affinity regression test
Changes to the SLURM configuration regarding the CPU affinity could affect the expected performance of applications. We should add a regression test that checks the processor affinity. I suggest using my tool [here](https://github.com/vkarak/affinity) for that.

Internal issue: https://madra.cscs.ch/scs/reframe/issues/150

### PR DESCRIPTION
**Pull request by rsarm**
Closes #100

### CODE REVIEW

**Comment by jgphpc** [comment]
For the sake of clarity, i would add (the output of `hwloc`) for daint-gpu and daint-mc to 
`gpu_omp_bind_cores.txt` and `mc_omp_bind_cores.txt` (just for info):

## daint-gpu

```
| | /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\ | |
| | | Core P#0    |  | Core P#1    |  | Core P#2    |  | Core P#3    |  | Core P#4    |  | Core P#5    |  | Core P#8    |  | Core P#9    |  | Core P#10   |  | Core P#11   |  | Core P#12   |  | Core P#13   | | |
| | |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             | | |
| | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | |
| | | | PU P#0  | |  | | PU P#1  | |  | | PU P#2  | |  | | PU P#3  | |  | | PU P#4  | |  | | PU P#5  | |  | | PU P#6  | |  | | PU P#7  | |  | | PU P#8  | |  | | PU P#9  | |  | | PU P#10 | |  | | PU P#11 | | | |
| | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | |
| | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | |
| | | | PU P#12 | |  | | PU P#13 | |  | | PU P#14 | |  | | PU P#15 | |  | | PU P#16 | |  | | PU P#17 | |  | | PU P#18 | |  | | PU P#19 | |  | | PU P#20 | |  | | PU P#21 | |  | | PU P#22 | |  | | PU P#23 | | | |
| | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | |
| | \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/ | |
| \------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------/ |
\----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------/
```

## daint-mc

```
| | | /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\
| | | | Core P#0    |  | Core P#1    |  | Core P#2    |  | Core P#3    |  | Core P#4    |  | Core P#8    |  | Core P#9    |  | Core P#10   |  | Core P#11   |
| | | |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |
| | | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |
| | | | | PU P#0  | |  | | PU P#1  | |  | | PU P#2  | |  | | PU P#3  | |  | | PU P#4  | |  | | PU P#5  | |  | | PU P#6  | |  | | PU P#7  | |  | | PU P#8  | |
| | | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |
| | | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |
| | | | | PU P#36 | |  | | PU P#37 | |  | | PU P#38 | |  | | PU P#39 | |  | | PU P#40 | |  | | PU P#41 | |  | | PU P#42 | |  | | PU P#43 | |  | | PU P#44 | |
| | | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |
| | | \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/
 /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\ | | |
 | Core P#16   |  | Core P#17   |  | Core P#18   |  | Core P#19   |  | Core P#20   |  | Core P#24   |  | Core P#25   |  | Core P#26   |  | Core P#27   | | | |
 |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             | | | |
 | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | | |
 | | PU P#9  | |  | | PU P#10 | |  | | PU P#11 | |  | | PU P#12 | |  | | PU P#13 | |  | | PU P#14 | |  | | PU P#15 | |  | | PU P#16 | |  | | PU P#17 | | | | |
 | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | | |
 | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | | |
 | | PU P#45 | |  | | PU P#46 | |  | | PU P#47 | |  | | PU P#48 | |  | | PU P#49 | |  | | PU P#50 | |  | | PU P#51 | |  | | PU P#52 | |  | | PU P#53 | | | | |
 | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | | |
 \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/ | | |
| \---------------------------------------------------------------------------------------------------------------------------------------------------------
| | | /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\
| | | | Core P#0    |  | Core P#1    |  | Core P#2    |  | Core P#3    |  | Core P#4    |  | Core P#8    |  | Core P#9    |  | Core P#10   |  | Core P#11   |
| | | |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |
| | | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |
| | | | | PU P#18 | |  | | PU P#19 | |  | | PU P#20 | |  | | PU P#21 | |  | | PU P#22 | |  | | PU P#23 | |  | | PU P#24 | |  | | PU P#25 | |  | | PU P#26 | |
| | | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |
| | | | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |
| | | | | PU P#54 | |  | | PU P#55 | |  | | PU P#56 | |  | | PU P#57 | |  | | PU P#58 | |  | | PU P#59 | |  | | PU P#60 | |  | | PU P#61 | |  | | PU P#62 | |
| | | | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |
| | | \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/
 /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\  /-------------\ | | |
 | Core P#16   |  | Core P#17   |  | Core P#18   |  | Core P#19   |  | Core P#20   |  | Core P#24   |  | Core P#25   |  | Core P#26   |  | Core P#27   | | | |
 |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             |  |             | | | |
 | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | | |
 | | PU P#27 | |  | | PU P#28 | |  | | PU P#29 | |  | | PU P#30 | |  | | PU P#31 | |  | | PU P#32 | |  | | PU P#33 | |  | | PU P#34 | |  | | PU P#35 | | | | |
 | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | | |
 | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ |  | /---------\ | | | |
 | | PU P#63 | |  | | PU P#64 | |  | | PU P#65 | |  | | PU P#66 | |  | | PU P#67 | |  | | PU P#68 | |  | | PU P#69 | |  | | PU P#70 | |  | | PU P#71 | | | | |
 | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ |  | \---------/ | | | |
 \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/  \-------------/ | | |
```

**Comment by jgphpc** [comment]
For info, there are variations between compilers on daint-mc (-n1 -c36, no multithreading):
screenshot 2019-01-25 at 14 06 36

**Comment by vkarak** [comment]
@jenkins-cscs retry daint

File: cscs-checks/prgenv/affinity_check.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        self.maintainers = ['RS', 'VK']
+        self.tags = {'production', 'scs'}
+
+    def setup(self, partition, environ, **job_opts):
+
+        def parse_affinity(x):
+            x_int = [int(xi) for xi in x.split()]
+            return sorted(x_int)
+
+        re_aff_cores = r'CPU affinity: \[\s+(?P<cpu>[\d+\s+]+)\]'
```
Comment (vkarak) [inline cscs-checks/prgenv/affinity_check.py]:
```suggestion
        re_aff_cores = r'CPU affinity: \[\s+(?P<cpus>[\d+\s+]+)\]'
```

Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: I know that you wanted to save some space and be more compact, but I could not understand the name of the variable from the beginning. I would vote for `core_affinity_regex` as variable name. Am I being too picky?

The same applies to the other cases.

File: cscs-checks/prgenv/affinity_check.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            re_aff_thrds, self.cases[self.variant]['ref_%s' % partition.name],
+            'thread', int)
+        re_aff_ranks = r'^Tag:[^\n\r]*Rank:\s+(?P<rank>\d+)[\s+\S+]'
+        self.aff_ranks = sn.extractall(re_aff_ranks, self.stdout, 'rank', int)
+        self.ref_ranks = sn.extractall(
+            re_aff_ranks, self.cases[self.variant]['ref_%s' % partition.name],
+            'rank', int)
+
+        self.use_multithreading = self.cases[self.variant]['multithreading']
+
```
Comment (vkarak) [inline cscs-checks/prgenv/affinity_check.py]: It is a bit hard to following what is going here with the sanity checking. What I have understood is that you assume that ranks and threads will be printed in order, so later you can compare the individual sequences, i.e., the sequence of threads, ranks and the corresponding cpu lists. Am I right? Please write a comment here explaining what you are doing, because I believe that there is an assumption here that is not so obvious to follow.

File: cscs-checks/prgenv/affinity_check.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        re_aff_cores = r'CPU affinity: \[\s+(?P<cpu>[\d+\s+]+)\]'
+        self.aff_cores = sn.extractall(
+            re_aff_cores, self.stdout, 'cpu', parse_affinity)
+        self.ref_cores = sn.extractall(
+            re_aff_cores, self.cases[self.variant]['ref_%s' % partition.name],
+            'cpu', parse_affinity)
+        re_aff_thrds = r'^Tag:[^\n\r]*Thread:\s+(?P<thread>\d+)'
+        self.aff_thrds = sn.extractall(re_aff_thrds, self.stdout, 'thread',
+                                       int)
+        self.ref_thrds = sn.extractall(
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: I know that you wanted to save some space and be more compact, but I could not understand the name of the variable from the beginning. I would vote for ref_num_threads as variable name. Am I being too picky?

The same applies to the other cases.

File: cscs-checks/prgenv/affinity_check.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                'ntasks-per-core': 2,
+                'multithreading': None,
+                'OMP_PLACES': 'threads',
+            },
+            'omp_bind_threads_nomultithread': {
+                'ref_gpu': 'gpu_omp_bind_threads_nomultithread.txt',
+                'ref_mc': 'mc_omp_bind_threads_nomultithread.txt',
+                'num_cpus_per_task': 12,
+                'ntasks-per-core': None,
+                # with --hint = nomultithread not explicetly expecified,
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: I would write:
> when --hint=nomultithread is not expecified, only half of the cores is used

File: cscs-checks/prgenv/affinity_check.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                self.cases[self.variant]['num_cpus_per_task'])
+        else:
+            self.num_cpus_per_task = (
+                self.cases[self.variant]['num_cpus_per_task'] * 3)
+
+        self.num_tasks = 1
+        self.variables  = {
+            'OMP_NUM_THREADS': str(self.num_cpus_per_task),
+            'OMP_PLACES': self.cases[self.variant]['OMP_PLACES']
+            # OMP_PROC_BIND is set to TRUE if OMP_PLACES is defined
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: Sorry, did not understand the comment
How come that 
> OMP_PROC_BIND values MASTER, CLOSE and SPREAD give the same result as OMP_PROC_BIND=TRUE

File: cscs-checks/prgenv/affinity_check.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            },
+        }
+        self.num_tasks = self.cases[variant]['num_tasks']
+        self.num_cpus_per_task = self.cases[variant]['num_cpus_per_task']
+        self.num_tasks_per_socket = self.cases[variant]['num_tasks_per_socket']
+
+    def setup(self, partition, environ, **job_opts):
+        super().setup(partition, environ, **job_opts)
+        if self.cases[self.variant]['cpu-bind']:
+            self.job.launcher.options = ['--cpu-bind=%s' %
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: Should we check if the scheduler is SLURM here? 

We should either clean the programming environments if the scheduler is not SLURM (skip the test) or fail immediately. @vkarak, what do you think?

File: cscs-checks/prgenv/affinity_check.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        super().__init__(variant)
+        self.descr = 'Checking distribution of ranks and threads over sockets.'
+        self.valid_systems = ['daint:mc', 'dom:mc']
+        self.cases = {
+            'alternate_socket_filling': {
+                'ref_mc': 'alternate_socket_filling.txt',
+                'num_tasks': 36,
+                'num_cpus_per_task': 1,
+                'num_tasks_per_socket': 18,
+                'multithreading': False,
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: There is no `'multithreading': True` or `'multithreading': None` test cases for `SocketDistributionTest`. Why?

File: cscs-checks/prgenv/affinity_check.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            sn.assert_eq(self.aff_ranks, self.ref_ranks),
+            sn.assert_eq(self.aff_cores, self.ref_cores)])
+
+        super().setup(partition, environ, **job_opts)
+
+
+@rfm.parameterized_test(['omp_bind_threads'],
+                        ['omp_bind_threads_nomultithread'],
+                        ['omp_bind_cores'])
+class AffinityOpenMPTest(AffinityTestBase):
```
Comment (victorusu) [inline cscs-checks/prgenv/affinity_check.py]: Don't you have to have 
```
self.build_system.options += ['OPENMP=1']
```
for this case? You are testing OpenMP affinity, but the code is not OpenMP enabled.

File: cscs-checks/prgenv/src/gpu_omp_bind_cores.txt [hunk 9]
```diff
@@ -0,0 +1,24 @@
+Tag: affinity/affinity, Hostname: nid00001, Rank: 0, Thread: 0
```
Comment (victorusu) [inline cscs-checks/prgenv/src/gpu_omp_bind_cores.txt]: How come you have several threads, if the code is not OpenMP enabled?

**Comment by vkarak** [comment]
@jgphpc You have a point with your comment. However, I wouldn't block this PR for this. I would rather create a follow up one for fixing what you suggest.

File: cscs-checks/prgenv/affinity_check.py (hunk 1 above)
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: @victorusu I didn't have problem understanding the names. To my opinion, what is missing is a bit of explanation about the logic as I pointed out in my review.

**Comment by jgphpc** [comment]
that was my goal: we need to open a new issue after this one.

File: cscs-checks/prgenv/affinity_check.py (hunk 3 above)
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: @victorusu I don't prefer very long names.

File: cscs-checks/prgenv/affinity_check.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+class AffinityTestBase(rfm.RegressionTest):
+    def __init__(self, variant):
+        super().__init__()
+        self.valid_systems = ['daint:gpu', 'daint:mc',
+                              'dom:gpu', 'dom:mc']
+        self.valid_prog_environs = ['PrgEnv-gnu']
+        self.build_system = 'Make'
+        self.build_system.options = ['--directory=affinity', 'MPI=1']
+        self.prebuild_cmd = ['git clone https://github.com/vkarak/affinity']
```
[the remark this answers is not in the document]
Comment (rsarm) [reply cscs-checks/prgenv/affinity_check.py]: I did that when I started writing the test, but then I realized that when setting `sourcesdir` to the repository URL, the `src` folder is not copied on `stage` and then the reference files are not visible by ReFrame.

File: cscs-checks/prgenv/affinity_check.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        self.variant = variant
+
+    def setup(self, partition, environ, **job_opts):
+        if partition.name == 'gpu':
+            self.num_cpus_per_task = (
+                self.cases[self.variant]['num_cpus_per_task'])
+        else:
+            self.num_cpus_per_task = (
+                self.cases[self.variant]['num_cpus_per_task'] * 3)
```
[the remark this answers is not in the document]
Comment (rsarm) [reply cscs-checks/prgenv/affinity_check.py]: This is to fill all threads on the `mc` partition.

File: cscs-checks/prgenv/affinity_check.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: Can you add a comment before this line stating exactly this?

File: cscs-checks/prgenv/affinity_check.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: Can you comment in the code?

File: cscs-checks/prgenv/affinity_check.py (hunk 8 above)
Comment (rsarm) [reply cscs-checks/prgenv/affinity_check.py]: `OPENMP` is 1 by default on the Makefile.

File: cscs-checks/prgenv/src/gpu_omp_bind_cores.txt (hunk 9 above)
Comment (rsarm) [reply cscs-checks/prgenv/src/gpu_omp_bind_cores.txt]: Same as above, `OPENMP ?= 1 ` on the Makefile.

File: cscs-checks/prgenv/affinity_check.py (hunk 6 above)
Comment (rsarm) [reply cscs-checks/prgenv/affinity_check.py]: Is it checking that `self.job.launcher.registered_name == srun` enough to say that the scheduler is Slurm?

File: cscs-checks/prgenv/affinity_check.py (hunk 7 above)
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: @victorusu We are not checking all different combination of flags here. This would be impractical. The purpose of the test is to use the necessary flags and environment variables in order to achieve specific and meaningful rank and thread placements.

**Comment by vkarak** [comment]
@victorusu Can you check and approve this PR?

File: cscs-checks/prgenv/affinity_check.py (hunk 6 above)
Comment (vkarak) [reply cscs-checks/prgenv/affinity_check.py]: @victorusu  Since currently the test is valid only for `daint/dom`, I think it's fine to do this assumption.
